File: src/main/java/engine/feature/shader/DefaultShaderCreator.java
package engine.feature.shader;

import com.jogamp.opengl.GL4;
import org.jetbrains.annotations.NotNull;

import java.io.IOException;
import java.net.URL;

public class DefaultShaderCreator implements ShaderCreator {

    public DefaultShaderCreator() {
    }

    @NotNull
    private String getShaderSource(@NotNull String resourceName) {
        URL fileURL = getClass().getClassLoader().getResource(resourceName);
        String out = null;
        try {
            out = Shader.readFromFile(fileURL.getPath());
        } catch (IOException e) {
            e.printStackTrace();
        }

        return out;
    }

    @NotNull
    private String[] getShaderBuffer(@NotNull String resourceName) {
        return new String[]{getShaderSource(resourceName)};
    }

    @NotNull
    @Override
    public Shader create(@NotNull String vertexResName, @NotNull String fragmentResName, @NotNull GL4 gl) {
        Shader out = new Shader(gl);
        out.compile(getShaderBuffer(vertexResName), getShaderBuffer(fragmentResName), null);

        return out;
    }

    @NotNull
    @Override
    public Shader create(@NotNull String vertexResName, @NotNull String fragmentResName, @NotNull String geometryResName, @NotNull GL4 gl) {
        Shader out = new Shader(gl);
        out.compile(getShaderBuffer(vertexResName), getShaderBuffer(fragmentResName), getShaderBuffer(geometryResName));

        return out;
    }
}

File: src/main/java/engine/feature/shader/DefaultShaderCreator.kt
package engine.feature.shader

import com.jogamp.opengl.GL4
import engine.feature.shader.`interface`.ShaderCreator

import java.io.IOException

class DefaultShaderCreator : ShaderCreator {

    private var gl: GL4? = null

    private fun getShaderSource(resourceName: String): String {
        val fileURL = javaClass.classLoader.getResource(resourceName)
        var out = ""

        try {
            out = Shader.readFromFile(fileURL!!.path)
        } catch (e: IOException) {
            e.printStackTrace()
        }

        return out
    }

    private fun getShaderBuffer(resourceName: String): Array<String> = arrayOf(getShaderSource(resourceName))

    override fun attachGl(gl: GL4) {
        this.gl = gl
    }

    override fun create(vertexSource: String, fragmentSource: String): Shader {
        val out = Shader(gl)
        out.compile(getShaderBuffer(vertexSource),
                    getShaderBuffer(fragmentSource),
                    null)

        return out
    }

    override fun create(vertexSource: String, fragmentSource: String, geometrySource: String): Shader {
        val out = Shader(gl)
        out.compile(getShaderBuffer(vertexSource),
                    getShaderBuffer(fragmentSource),
                    getShaderBuffer(geometrySource))

        return out
    }
}
